File: solutions/459_repeated_substring_pattern.py
"""
Given a non-empty string check if it can be constructed by taking a substring of it and appending multiple copies of the substring together.  You may assume the given string consists of lowercase English letters only and its length  will not exceed 10000. 
Example 1:
Input: "abab"
Output: True
Explanation: It's the substring "ab" twice.
Example 2:
Input: "aba"
Output: False
Example 3:
Input: "abcabcabcabc"
Output: True
Explanation: It's the substring "abc" four times. (And the substring "abcabc" twice.)
"""


class Solution:
    def repeatedSubstringPattern(self, s):
        """
        :type s: str
        :rtype: bool
        """
        i = 1
        true, false = [True, False]
        while i <= len(s) / 2:
            sub = s[0:i]
            j = i
            k = i + i
            while k <= len(s) and sub == s[j:k]:
                j += i
                k += i
            if j == len(s): return true
            i += 1
        return false

    def repeatedSubstringPattern2(self, s):
        return any(s[:i] * int(len(s) / i) == s
                   for i in range(1,
                                  int(len(s) / 2) + 1) if len(s) % i == 0)

    def repeatedSubstringPattern3(self, s):
        return s in (s * 2)[1:-1]


s = "abcabcabcabc"
s = "abaababaab"
print(Solution().repeatedSubstringPattern(s))
print(Solution().repeatedSubstringPattern2(s))
print(Solution().repeatedSubstringPattern3(s))

print(((s + s)[1:-1]))

File: _site/python_solutions/459.repeated-substring-pattern.rb
#
# @lc app=leetcode id=459 lang=ruby
#
# [459] Repeated Substring Pattern
#
# https://leetcode.com/problems/repeated-substring-pattern/description/
#
# algorithms
# Easy (39.06%)
# Total Accepted:    70.2K
# Total Submissions: 179.4K
# Testcase Example:  '"abab"'
#
# Given a non-empty string check if it can be constructed by taking a substring
# of it and appending multiple copies of the substring together. You may assume
# the given string consists of lowercase English letters only and its length
# will not exceed 10000.
#
#
#
# Example 1:
#
#
# Input: "abab"
# Output: True
# Explanation: It's the substring "ab" twice.
#
#
# Example 2:
#
#
# Input: "aba"
# Output: False
#
#
# Example 3:
#
#
# Input: "abcabcabcabc"
# Output: True
# Explanation: It's the substring "abc" four times. (And the substring "abcabc"
# twice.)
#
#
#
# @param {String} s
# @return {Boolean}
def repeated_substring_pattern(s)
  return false if s.length <= 1

  s.each_char.with_index do |c, i|
    next if i.zero?
    return true if c == s[0] && s[0..i - 1] * (s.size / i) == s
  end
  false
end

s = 'abab'
s = 'aba'
s = 'abcabcabcabc'
p repeated_substring_pattern(s)
